### PR TITLE
Fix semibatch reactor energy flow term

### LINKED ISSUE #87: High: [Reactors] Semibatch energy balance subtracts fictitious outflow enthalpy
## Maintainer triage (June 29, 2026)
- **Status:** Verified bug against the current `master` source in this fork.
- **Severity:** High.
- **Area labels:** area:reactors.
- **Tackle batch:** P2 - shared thermo, phase, and basis correctness.

The source review confirmed that the cited code path and failure mode are still present. The original audit details are preserved below for reproduction notes and suggested fixes.

---

## Original audit details

**Severity:** High · Part of #67 (round-2 audit batch)

### Convective enthalpy term weights the second enthalpy by holdup instead of inlet concentration, contradicting the no-outflow material balance
**Location:** `PharmaPy/Reactors.py:1046-1048 (CSTR.energy_balances flow_term), reached for SemibatchReactor via lines 1289-1294 and 1013-1069`

SemibatchReactor reuses CSTR.energy_balances. flow_term = inlet_flow*(h_in - h_temp) with h_in = sum(inlet_conc*h_inj) (inlet composition) and h_temp = sum(mole_conc*h_tempj) (HOLDUP composition) (lines 1046-1048). This is the correct in-minus-out form for a CSTR, but a semibatch reactor has NO outflow: its material balance (1289-1294, dN_i/dt = F*C_in,i + r_i*V; dvol_dt = inlet_flow at 1292) is pure accumulation. Deriving the consistent temperature-form balance gives NCp*dT/dt = F*sum(C_in,i*(h_in,i(T_in) - h_i(T))) + source - ht, i.e. BOTH enthalpy terms must be weighted by inlet C_in,i. The code instead subtracts F*sum(C_i*h_i(T)) using holdup C_i, an enthalpy of a nonexistent outflow, leaving a spurious error term F*sum((C_in,i - C_i)*h_i(T)). Since h_i is sensible enthalpy referenced to 298.15 K, the error does not vanish even when T_in ~ T.

- **Trigger:** SemibatchReactor(vol_tank=..., isothermal=False) (isothermal=False is documented, lines 1233-1235) with a feed composition different from the initial charge (the normal case), then solve_unit(). Exercised in tests/Flowsheet/flowsheet_tests.py:131 (holdup B+solvent, feed pure A).
- **Wrong result:** dtemp_dt (line 1069) uses the wrong convective enthalpy, so the temperature trajectory is wrong; through Arrhenius temp_term this also corrupts reaction rates and hence the concentration profiles of every non-isothermal semibatch run with feed composition != charge. No crash; silently wrong.
- **Suggested fix:** Override energy_balances for SemibatchReactor (or branch in CSTR.energy_balances on 'vol' in states_uo): replace h_temp with the inlet-weighted holdup enthalpy, flow_term = inlet_flow*(inlet_conc*(h_inj - h_tempj)).sum(axis=1)*1000 = F*sum(C_in,i*(h_in,i(T_in) - h_i(T))), removing the spurious outflow-at-holdup-composition term.
- **Net-new (not a duplicate):** Not #34 (CSTR getHeatOfRxn signature) nor #38/#25/#24/#48/#68/#80 (enthalpy-reference issues in other units). No filed item addresses the semibatch convective enthalpy basis / material-energy inconsistency.

**Related existing issues (referenced, not modified):** #34 (CSTR equilibrium getHeatOfRxn signature), #38/#25/#24/#48/#68/#80 (Mixer/crystallizer/drying/extraction enthalpy-reference issues) -- reference only; none address the semibatch convective enthalpy term.

---
_Found by a multi-agent source audit (round 2); confirmed by re-reading the source and cross-checked against all 80 existing open issues. Each defect was verified by an independent code-truth skeptic and a novelty skeptic._

### PR DESCRIPTION
**Pull request by parkyr**
## Summary

- Correct the semibatch reactor convective energy term so sensible enthalpy at reactor temperature is weighted by inlet composition, matching the semibatch material accumulation balance with no outlet stream.
- Preserve CSTR behavior by keeping the holdup-composition outflow enthalpy for reactors without the semibatch `vol` state.
- Add algebra-level regression tests for the semibatch zero-driving-force case and the CSTR outflow guard.

Closes #87.

## Related Issues / Priority

- The Assimulo import workaround in the new algebra-level test is temporary and directly related to #10, which tracks lazy solver imports plus a solver interface / SciPy fallback.
- #10 should be prioritized before broad solver-adjacent test expansion or larger unit-operation refactors, because it removes the need for import-time stubs in core tests.
- #10 does not need to block this PR: this PR is a focused correctness fix for #87, keeps CSTR behavior pinned, and passes both core and Assimulo CI.
- Related roadmap context: #3 covers the broader CI / engineering baseline, #7 covers test expansion, and #8 covers packaging metadata cleanup.

## Tests

- `conda run -n pharmapy python -m pytest tests/test_reactors_energy.py -q`
- `conda run -n pharmapy python -m pytest --collect-only`
- `conda run -n pharmapy python -m pytest tests/ -m "not assimulo"`
- `timeout 180s conda run -n pharmapy python -m pytest tests/Flowsheet/flowsheet_tests.py::TestFlowsheets::test_SB_BC_FILT -q`
- `timeout 180s conda run -n pharmapy python -m pytest tests/ -m assimulo -q`
- `git -c core.whitespace=blank-at-eol,blank-at-eof,space-before-tab,cr-at-eol diff --check --cached`

Notes: pytest emitted existing invalid-escape deprecation warnings in several modules and one existing runtime warning in the PFR flowsheet test.

## Branch Hygiene

- Base branch: `master`
- Source branch: `fix/issue-87-semibatch-energy`
- Branch point: `origin/master` at `1c3a046fceb5c4da8b5fdb6e8b254f657d6e3cc7`
- Upstream status: `origin/master` includes `upstream/master` at `400e1e13ad4ecd3ba202d5fae8c70ec73bdf9c3e`
- Stacked status: not stacked; no prerequisite PRs

### CODE REVIEW

**Review by parkyr — COMMENTED** [review]
## Maintainer review — Fix semibatch reactor energy flow term (#102, closes #87)

Reviewed head `f8dad35` against base `1c3a046`. The fix is correct, minimal, and well-scoped. Posting as **COMMENT** because I authored this PR (GitHub rejects an author `APPROVE`).

### Correctness — verified
The change branches the convective second-enthalpy term on `'vol' in self.states_uo`, weighting `h_temp` by inlet composition for the semibatch (no outlet stream) and keeping holdup composition for the CSTR outflow. I independently confirmed:
- `SemibatchReactor.nomenclature` is the only path that appends `'vol'` to `states_uo`, and it runs during `__init__` (via `CSTR.__init__ → self.nomenclature()`), so the guard is `True` for a semibatch and `False` for CSTR/Batch/PFR both at construction and in the real solve. The fix also reuses the exact idiom already present at `Reactors.py:1084`.
- Regression tests are **not vacuous**: run against true base source (`1c3a046`), `test_semibatch_flow_term_...` fails with `-1.037e7 ≠ 0` and passes on head; `test_cstr_flow_term_keeps_holdup_composition_for_outflow` asserts `-1.037e7` and passes on both base and head, so it locks in "no CSTR regression" (issue #87 acceptance criterion #3).

### Tests run locally (env `pharmapy`, assimulo present)
- `tests/test_reactors_energy.py` — 2 passed (head); semibatch case fails on base as expected.
- `pytest tests/ -m "not assimulo"` — 29 passed, 6 deselected.
- `test_SB_BC_FILT` (assimulo integration) — passed.
- CI on this head: **Core tests** SUCCESS, **Assimulo integration tests** SUCCESS.

Warnings observed are pre-existing invalid-escape `DeprecationWarning`s in unrelated modules; not introduced here.

### Nonblocking

`Nonblocking`: The integration test `test_SB_BC_FILT` still only asserts the solve completes (no assertion on the corrected temperature trajectory), so the full-solve path would not catch a future regression of this term — the new algebra tests are what guard it. Consider adding a lightweight assertion there (e.g. an energy-balance/temperature check or a golden trace) so the integration path also protects the fix. The unit-level coverage already makes this optional. 

`Nonblocking`: `tests/test_reactors_energy.py` stubs `assimulo` and rewrites `sys.path`/`sys.modules` (`_prefer_source_package`, `_stub_assimulo_modules`) so the algebra test can import `PharmaPy.Reactors` in the no-assimulo Core job. This is justified (the top-level `from assimulo... import` would otherwise fail), but it diverges from the repo's existing pattern of guarding the import behind `if HAS_ASSIMULO:` (see `tests/integration/reactor_tests.py`). Fine to keep; noting for consistency. 

### Merge readiness
No blocking findings. `mergeStateStatus` is `CLEAN` and `reviewDecision` is empty (no required-review gate reported by branch protection). This COMMENT does not constitute a formal approval, and this account cannot self-approve its own PR; if an approval is required before merge, an eligible maintainer account will need to submit it.

**Comment by parkyr** [comment]
Adding a quick note to connect this test workaround with the PR body's **Related Issues / Priority** section.

The Assimulo monkeypatch in this test is temporary glue, not the desired long-term design. Right now, `PharmaPy.Reactors` imports Assimulo at module import time, so even simple algebra-only unit tests can fail when Assimulo/SUNDIALS is not installed.

That broader problem is already tracked in #10, which proposes lazy solver imports plus a solver interface / SciPy fallback. I think #10 should be prioritized before broad solver-adjacent test expansion or larger unit-operation refactors, because it would remove the need for import-time stubs in core tests.

That said, #10 does not need to block this PR. This PR is a focused correctness fix for #87, and the workaround is only here so the regression test can run under the current package structure.

Related roadmap context: #3 is the broader CI / engineering baseline umbrella, #7 covers test expansion, and #8 covers packaging / dependency cleanup.

File: PharmaPy/Reactors.py [hunk 1]
```diff
@@ -1044,7 +1044,12 @@ def energy_balances(self, time, mole_conc, vol, temp, temp_ht, inputs,
                                    total_h=False, basis='mole')
 
         h_in = (inlet_conc * h_inj).sum(axis=1) * 1000  # J/m**3
-        h_temp = (mole_conc * h_tempj).sum(axis=1) * 1000  # J/m**3
+
+        if 'vol' in self.states_uo:  # Semibatch: no outlet enthalpy stream
+            h_temp = (inlet_conc * h_tempj).sum(axis=1) * 1000  # J/m**3
```
Comment (bernalde) [inline PharmaPy/Reactors.py]: Let’s have this be the chance that if we are adding units in the comments, we switch them to [unit] with the same units than the package pint. I would suggest, let’s open an issue for this but also while adding new code, let’s start pushing for this standard

**Review by bernalde — CHANGES_REQUESTED** [review]
An issue should be created and some commented units should be standarized

**Comment by bernalde** [comment]
Answering to the last comment, the order of the merging would affect the work here. We can merge this now, but once #10 is tackled we would need to come and revisit this code to fix it. Otherwise, we merge the solution to #10 and from there we resolve the merge conflicts here. Do as you wish, but just let me (and GitHub) know

**Review by andres9403 — APPROVED** [review]

**Comment by bernalde** [comment]
@parkyr What did you choose in terms of the merging Sequence (see my last comment)?

**Comment by parkyr** [comment]
> @parkyr What did you choose in terms of the merging Sequence (see my last comment)?

From what I understand, fixing #10 would only change the tests in this pr (which we are using as a workaround to test the fixes--essentially bypassing the issue #10 would fix). And these tests would still work regardless of #10 fix to test the bug targeted in this PR. Also, considering #10 would likely affect many tests/codes, I think that would likely end up being an epic issue on its own, which should be tackled similarly to the documentation issue perhaps?
 
In short, I think it'd be best to have this pr merged (fixing the flow term), and the assimulo fix should be tackled separately after #10 fix.

**Comment by parkyr** [comment]
## PR Summary Update

- This update captures review follow-up and merge-order notes without replacing the original PR summary.
- Addressed the units-comment review request in `ca6a810`: the touched reactor energy-balance comments now use bracketed units (`[J/m**3]`, `[W]`).
- Broader units and basis conventions remain tracked in #133; this PR stays focused on the semibatch energy-balance fix for #87.
- Recommended merge order: let #135 land first, then refresh this branch on current `master`.
- After #135 lands, simplify `tests/test_reactors_energy.py` if `PharmaPy.Reactors` imports without the temporary fake Assimulo modules.

## Current Checks

- `conda run -n pharmapy python -m pytest tests/test_reactors_energy.py -q` - 2 passed.
- `git -c core.whitespace=blank-at-eol,blank-at-eof,space-before-tab,cr-at-eol diff --check` - passed.
- Current-head GitHub checks at `ca6a810` passed: Core tests, Assimulo integration tests, and locked pixi install on Ubuntu and Windows.

## Before Re-Review

- Rebase or merge current `master` after #135 lands.
- Rerun the focused reactor test and the relevant core/Assimulo lanes.
- Ask `bernalde` to re-review or dismiss the standing `CHANGES_REQUESTED`.
